db/types.go

```diff
@@ -0,0 +1,124 @@
+package db
+
+type DB interface {
+
+	// Get returns nil iff key doesn't exist. Panics on nil key.
+	Get([]byte) []byte
+
+	// Has checks if a key exists. Panics on nil key.
+	Has(key []byte) bool
+
+	// Set sets the key. Panics on nil key.
+	Set([]byte, []byte)
+	SetSync([]byte, []byte)
+
+	// Delete deletes the key. Panics on nil key.
+	Delete([]byte)
+	DeleteSync([]byte)
+
+	// Iterator over a domain of keys in ascending order. End is exclusive.
+	// Start must be less than end, or the Iterator is invalid.
+	// CONTRACT: No writes may happen within a domain while an iterator exists over it.
+	Iterator(start, end []byte) Iterator
+
+	// Iterator over a domain of keys in descending order. End is exclusive.
+	// Start must be greater than end, or the Iterator is invalid.
+	// CONTRACT: No writes may happen within a domain while an iterator exists over it.
+	ReverseIterator(start, end []byte) Iterator
+
+	// Releases the connection.
+	Close()
+
+	// Creates a batch for atomic updates.
+	NewBatch() Batch
+
+	// For debugging
+	Print()
```

Comment:
Personally I'd prefer
```go
Fprintf(w io.Writer)
```

Otherwise our stdout is about to get lit for every database and of which extending that to print to the screen during tests will just be a matter of `Fprint(os.Stdout)`. Using `Fprint` here all of a sudden gives us the power
to test our DBs easily, periodically print out output to a log file for consistency checks, send overview over a network etc.

Reply:
Lets make another PR for that.

Reply:
We shouldn't be using Print for consistency checks (it's explicitly for human consumption) but it makes sense to create an overview to send over the network (e.g. RPC endpoint).  I'd prefer Sprint() string though.

Reply:
That way we kinda implicitly tell the user that this isn't designed for large outputs., just for debugging.